File: twice-linear.py
from collections import deque

'''https://www.codewars.com/kata/5672682212c8ecf83e000050
4 kyu
'''

def dbl_linear(n):
    u = [1]
    Q2, Q3 = [3], [4]
    
    for _ in range(n):
        next_val = min(Q2[0], Q3[0])
        u.append(next_val)

        if next_val == Q2[0]:
            Q2.pop(0)
        if next_val == Q3[0]:
            Q3.pop(0)

        Q2.append(2 * next_val + 1)
        Q3.append(3 * next_val + 1)

    return u[n]

File: twice-linear.c
#include <stdio.h>
#include <stdlib.h>

/*https://www.codewars.com/kata/5672682212c8ecf83e000050
4 kyu
*/

int min(int a, int b) {
    return a < b ? a : b;
}

int dblLinear(int n) {
    int *u = (int *)malloc((n + 1) * sizeof(int));

    u[0] = 1;
    
    int *Q2 = (int *)malloc((n + 1) * sizeof(int));
    int *Q3 = (int *)malloc((n + 1) * sizeof(int));

    Q2[0] = 3;
    Q3[0] = 4;

    int q2_index = 0, q3_index = 0;
    int q2_size = 1, q3_size = 1;

    for (int i = 1; i <= n; i++) {
        int next_val = min(Q2[q2_index], Q3[q3_index]);
        u[i] = next_val;

        if (next_val == Q2[q2_index]) {
            q2_index++;
        }
        if (next_val == Q3[q3_index]) {
            q3_index++;
        }

        Q2[q2_size++] = 2 * next_val + 1;
        Q3[q3_size++] = 3 * next_val + 1;
    }

    free(u);
    free(Q2);
    free(Q3);

    return u[n];
}
